## Summary Bayesian Alphabet

**\--sbayes** R

Available models are R, S and C.

**\--ldm** test.ldm.sparse

Input the LD matrix generated by **--make-sparse-ldm** option. This option tells GCTB to read two files: a text file that contains SNP information (test.ldm.sparse.info) and a binary file that contains the sparse LD matrix data (test.ldm.sparse.bin). The suffix ".sparse" tells GCTB the format of the LD matrix. The valid suffixes include ".full", ".band", ".shrunk" and ".sparse", which are generated from the corresponding options that make the LD matrix in sush a format, e.g. **--make-full-ldm** for ".full" files. 

**\--mldm** mldm.list

Input a list of LD matrices, e.g. from chromosome 1 to 22. The LD matrices need to be in the correct order in mldm.list.


**\--make-full-ldm**

Compute a full LD matrix from the genotype data provided by **--bfile** option. This option will generate two files: a text file that contains SNP information (test.ldm.full.info) and a binary file that contains the full LD matrix data (test.ldm.full.bin). The full LD matrix contains all pairwise LD correlations between SNPs. The memory required is approximately the size of the n-by-m genotype matrix (4 * n * m)/1e9 GB + the size of the m-by-m LD matrix (4 * m^2)/1e9 GB (n is the number of individuals and m is the number of SNPs). For a more efficient computation, it is recommended to use **--snp** option to compute the matrix by parts. 

**\--part** 5,1

Work only with --make-full-ldm. Compute a full LD matrix part by part to save memory. The first number before comma indicates the total parts to divide, the second number indicates the current part to calculate. Run gctb with the --mldm output.mldm to merge, after all parts are calculated successfully. 

**\--make-sparse-ldm**

Compute a sparse LD matrix from the genotype data provided by **--bfile** option or from the dense type LD matrix, such as full or shrunk LD matrix provided by **--ldm test.ldm.full** or **--ldm test.ldm.shrunk** option. If the genotype data is provided, GCTB will first compute the chromosome-wide full LD matrix, then set the chance LD (small LD correlations due to sampling) to be zero and output the sparse matrix. The chance LD is detected by a chi-squared test with the threshold defined by **--chisq**. The chi-squared threshold value is 10 by default. The memory usage is the same as computing a full LD matrix.

**\--make-shrunk-ldm**

Compute a shrunk LD matrix from the genotype data provided by **--bfile** option or from the full LD matrix provided by **--ldm test.ldm.full** option. It will use the method proposed by [Wen and Stephens (2010)](https://projecteuclid.org/euclid.aoas/1287409368) to shrunk the off-diagonal entries of the sample LD matrix toward zero based on a genetic map provided by **--gen-map** option (see [Tutorial](https://cnsgenomics.com/software/gctb/#Tutorial) for the format of the genetic map). Other parameters required by the shrinkage method include the effective population size provided by **--ne** option (default value is 11400), the genetic map sample size provided by **--genmap-n** option (default value is 183) and the shrinkage cutoff threshold provided by **--shrunk-cutoff** option (default value is 1e-5). The output format of the shrunk LD matrix is ".shrunk", which contains all the zero elements, so it will take a similar storage space as the full LD matrix. We recommend to further make the shrunk LD matrix parse by eliminating elements equal to zero for efficient storage and computation, which can be done by reading the generated shrunk matrix and making it sparse using **--make-sparse-ldm --chisq 0** options. The memory usage is the same as computing a full LD matrix.

**\--make-band-ldm**

Compute a banded LD matrix from the genotype data provided by **--bfile** option or from the full LD matrix provided by **--ldm test.ldm.full** option. GCTB will only compute and include the LD correlations between all paris of SNPs within a sliding window which has a window width defined by **--wind** option (in the unit of megabase). 

**\--snp** 1-5000

Compute the LD matrix by rows (or columns as it is a symmetric square matrix). When this option is invoked, GCTB will read in the genotype matrix from the bed file and compute a part of the LD matrix corresponding to the range of SNPs specified. GCTB will append `.snp1-5000`, for example, to the output files. Once all parts have been computed individually, make sure to include all the files as a list in the ascending order as a separate file and use **--mldm** option to merge the parts into a single matrix. See Toturial for more detail.

**\--gwas-summary** test.ma

Input the GWAS summary statistics in the [GCTA-COJO](https://cnsgenomics.com/software/gcta/#COJO) `.ma` format.

```
SNP A1 A2 freq b se p N 
rs1001 A G 0.8493 0.0024 0.0055 0.6653 129850 
rs1002 C G 0.0306 0.0034 0.0115 0.7659 129799 
rs1003 A C 0.5128 0.0045 0.0038 0.2319 129830
```

Columns are SNP identifier, the effect allele, the other allele, frequency of the effect allele, effect size, standard error, p-value 
and sample size. The headers are not keywords and will be omitted by the program. Important: "A1" needs to be the effect allele with 
"A2" being the other allele and "freq" should be the frequency of "A1". See [Tutorial](https://cnsgenomics.com/software/gctb/#Tutorial) for more information. 

**\--chisq** 10

Specify the chi-squared test threshold for computing the sparse LD matrix. The LD correlations that are less than the threshold (default value is 10) is regarded due to sampling variation and therefore removed from the LD matrix. It is equivalent to setting the LD rsq smaller than (chi-squared threshold)/(LD reference sample size).

**\--ld** 0.01

Specify the absolute value of LD correlation threshold for computing the sparse LD matrix. Any absolute value of the LD correlation below the threshold will be set to be zero. It cannot be used together with **\--chisq**.

**\--write-ldm-txt**

Output the LD matrix in a text file. 

**\--annot** annotation.txt

Input the genomic annotation data and launch the stratified SBayes analysis (SBayesS-strat). The format of the annotation file is as below.

```
SNP Coding Conserved DHS Promoter
rs1001 0 1 0 0 0
rs1002 1 0 0 0 0
rs1003 0 0 0 1 1
```

The first column is the SNP identifier, followed by annotation categories. The first row is the header line that contains the names of the annotation categories. The current version only allows binary annotations (i.e. 0/1) and only available for SBayesS model (i.e. together with **--sbayes S** option), which allows the SNP-heritability, polygenicity and the S parameter (relationship between effect size and MAF) to be different across annotation categories.

**\--exclude-mhc**

This option will remove SNPs in the MHC regions (chr6:28-34 Mb) from the analysis.

**\--ambiguous-snp**

This option will remove SNPs with ambiguous nucleotides, i.e. A/T or G/C.

**\--maf** 0.01

This option will remove SNPs with minor allele frequency from the GWAS sample smaller than the specified value.

**\--overlap**

This option tells GCTB that the LD reference sample is in complete overlap with the GWAS sample. That is, the LD matrix is computed from the GWAS sample itself. This will suppress the modeling of sampling variance of LD correlations in the Sbayes analysis.

**\--impute-n**

This option will impute per-SNP sample size in GWAS based on the GWAS effect estimates, SE and allele frequencies in the `.ma` file, and remove SNPs with the imputed sample size greater/smaller than 3 standard deviations from the median value.

**\--num-chains** 4

This option invokes a multi-chain MCMC process, e.g. 4 chains. GCTB will run multiple chains in sequence each with the starting values of the model parameters randomly sampled from their prior distributions. When the MCMC chains are completed, it will calculate the Gelman-Rubin convergence diagnostic (also known as potential scale reduction factor) for each of the model parameters and report them in the `.parRes` file. For example,

```
Posterior statistics from multiple chains:

            Mean            SD              R_GelmanRubin 
      Pi    0.044129        0.001628        1.005       
  NnzSnp    49875.335938    1825.153442     1.005       
 SigmaSq    0.000009        0.000000        1.007       
       S    -0.545913       0.029914        1.003       
  ResVar    0.727659        0.002897        1.000       
  GenVar    0.257742        0.002645        1.000       
SigmaSqG    0.258967        0.003865        1.000       
     hsq    0.261560        0.002507        1.000       
```

In the current verion, the multi-chain MCMC is only available for SBayesC, SBayesS and SBayesS-strat analysis.

**\--rsq** 0.9

For a pair of SNPs with the squared LD correlation greater than the specified value, e.g. 0.9, based on the provided LD matrix, this option will remove the SNP with the higher GWAS P value and keep the other SNP in the analysis.  

**\--p-value** 0.4

This option will remove SNPs with GWAS P value greater than the specified value, e.g. 0.4, from the analysis.

**\--robust**

This option will invoke the following parameterisation for the common factor of SNP effect variance in SBayesR, i.e., \\$\sigma\_{\beta}^2 = \frac{\sigma\_g^2 }{ m {\boldsymbol \gamma}'{\boldsymbol \pi} }\\$ where \\$\sigma\_g^2 = {\boldsymbol \beta} ' {\bf R} {\boldsymbol \beta}\\$ is the sampled total genetic variance in each MCMC iteration. Comparing to the original model that assumes \\$\sigma\_{\beta}^2\\$ following a scaled-inverse chi-square distribution, this parameterisation is observed to be more robust to the potential convergence issue which sometimes occurs when the GWAS summary statistics are from a meta-analysis. In GCTB 2.03 beta version, the program will start with the original model but if a convergence issue is detected (indicated by a negative sampled value of residual variance) it will restart the MCMC process with this more robust parameterisation. Note that this parameterisation is identical to LDpred ([Vilhj√°lmsson et al. 2015](https://www.sciencedirect.com/user/identity/landing?code=Ulj9Myn2m9Aa3R2j-wGhgPbv1WF9l5lxSIgHJR3u&state=retryCounter%3D0%26csrfToken%3D325c98ae-e6b3-43c0-b6ea-b6eba8ea096c%26idpPolicy%3Durn%253Acom%253Aelsevier%253Aidp%253Apolicy%253Aproduct%253Ainst_assoc%26returnUrl%3D%252Fscience%252Farticle%252Fpii%252FS0002929715003651%253Fvia%25253Dihub%26prompt%3Dnone%26cid%3Darp-d6cf6a05-621c-472f-b381-5809b9d8a27a)) when using a point-normal prior (i.e., two-component SBayesR or SBayesC).

**\--hsq-percentage-model true**

If you run SBayesR using GCTB v2.5.1 or above, the default model is the heritability percentage model, i.e., a mixture of normal distributions with the variance equal to \\$\gamma_k %\\$ of the estimated SNP-based heritability. To switch to the model described in the [SBayesR paper](https://www.nature.com/articles/s41467-019-12653-0), which considers the variance component \\$\sigma\_\beta^2\\$ as a free parameter, set the value to be 'false'.

**\--convert --snp-res test.snpRes --ldm-eigen ldm --extract snplist**

The **--convert** function maps the joint effect estimates of SNPs in the `test.snpRes` file to the joint effect estimates of a subset of SNPs defined in the `snplist` file, based on the LD information provided in the `ldm` folder. Since this mapping is done by LD blocks, multiple threads can be used to speed up the computation (e.g., **--thread 8**).

